### PR TITLE
Fix RewindBlockIndex for pruning

### PR DESCRIPTION
**Pull request by sdaftuar**
Turns out this is harder than I thought.  This is not suitable for merging because there's a bug (see below), and after further thought I now wonder whether trying to fix all this is worth it.

I ran into two issues: first, after we were calling `RewindBlockIndex`, in init we'd then call `VerifyDB`, which for a pruning node would fail because we just rewound back to the point where we have no more data, so trying to read the next earlier block from disk would fail.

I tried to workaround this by moving the `VerifyDB` call to happen before we call `RewindBlockIndex`, but then realized that this doesn't work, because eg if you were using checklevel=4, then the blocks that need to be rewound wouldn't in general pass `VerifyDB` (since we didn't download those blocks with the likely-needed witness data).

Maybe one solution to this problem would be to somehow bypass calling `VerifyDB` in the particular case that you're a pruning node and we just rewound the chain?  But that seems hacky.  Any better ideas?

The second issue I ran into had to do with wallet rescans taking place after `RewindBlockIndex`.  I figured there should be no problem, because the wallet ought to be ahead of the tip, which should be fine.  However, it turns out that `FindForkInGlobalIndex(chainActive, wallet-locator)` can return a blockindex that is earlier than `chainActive.Tip()` even if the wallet locator's tip is a successor of our tip.  So I added a quick "fix" to `FindForkInGlobalIndex` that would detect this situation.  It's not clear to me if there might be side effects (eg performance side effects, or new behavior because the function now returns different values in some situations) worth worrying about.

Also, I discovered that the node is in a pretty brittle state if it's a pruning node that is rewound this way...  If you were to, say, bring the node down after the chain has been rewound and before new blocks are downloaded from peers, then on restart the node will die in `VerifyDB`.  So at this point, I am wondering if we shouldn't perhaps just put a big warning in the release notes that pruning nodes should try to upgrade before activation in order to avoid redownloading the blockchain, along with graceful detection so that if the user fails to do so, they get a nicer error message?

I can take a stab at that approach if you agree that it makes more sense.

### CODE REVIEW

**Comment by sdaftuar** [comment]
Ok, I think this works now -- first commit changes the behavior of `-checkblocks` to not stomp on pruning nodes.  Hopefully the change to `FindForkInGlobalIndex` is okay...

File: src/main.cpp [hunk 1]
```diff
@@ -593,6 +593,9 @@ CBlockIndex* FindForkInGlobalIndex(const CChain& chain, const CBlockLocator& loc
             CBlockIndex* pindex = (*mi).second;
             if (chain.Contains(pindex))
                 return pindex;
+            if (pindex->GetAncestor(chain.Height()) == chain.Tip()) {
```
Comment (sdaftuar) [inline src/main.cpp]: Maybe this should only do the `GetAncestor` check once, on the first entry in the block locator?

**Comment by MarcoFalke** [comment]
(Note: The last commit is https://github.com/bitcoin/bitcoin/pull/8076 )

**Comment by sdaftuar** [comment]
Rebased (and got rid of unintended code move in `init.cpp` that was in the prior version).

**Comment by sipa** [comment]
Rebase on top of #8076?

**Comment by sdaftuar** [comment]
I cherry picked that commit from #8076 and rebased on sipa/segwit-master.

**Comment by sipa** [comment]
@sdaftuar Woops, I missed this is against segwit-master and not against upstream!

**Comment by sipa** [comment]
Including in next batch.

**Pull request closed without merging**